### ISSUE TITLE
Release Jep 3.9

### ISSUE THREAD

**Issue by bsteffensmeier** [ninia/jep#199]
It's been about a year since 3.8 and we have a few interesting new features so we would like to release 3.9. After discussing with @ndjensen here are the things we would like to get done before release. Ideally we would like to release in August or September 2019

- [x]  #144  change NamingConventionClassEnquirer to return itself
- [x]  #188 3b0910d update LibraryLocator or MainInterpreter to take explicit path.
- [x]  #168 6642ba7 Configuration of shared interpreters
- [x] Remove javax.script.engine stuff
- [x] Make a SubInterpreter class similar to Shared Interpreter. And deprecate Jep constructors for removal in 4.0.
- [x] Update Classlist for new java versions. At a minimum we need a classlist file for 11. We should consider adding a list for 12 or changing the code so it defaults to the list form the latest LTS relelase.
- [x] 4012182 Travis support for newer Python/Java
- [x] 93058da AppVeyor support for newer Python/Java
- [x] #198 93c1918 ensure we work on 3.8 beta releases. Consider trying python 3.9 too, should we add both to travis.yml?
- [x] #166 1756053 Fix `__name__` in SharedInterpreter
- [x] Update release notes

**Comment by ndjensen** [comment]
I updated the release notes.  They probably need some refining, and should mention something about the configuration of SharedInterpreters that is yet to be implemented.

**Comment by bsteffensmeier** [comment]
I would like to clean out all of the warnings from the Java compiler before we release 3.9 and it is exposing some larger design issues with interactive mode. I was planning on looking at this closer in Jep 4.0 but I don't think we can get rid of the warnings without introducing new API and I don't want to add something that would be immediately deprecated or removed in 4.0.

We are currently throwing a compilation warning because Jep.setInteractive() was [recently deprecated](https://github.com/ninia/jep/commit/025f231fd2bfe708c9bfaa868186fc6c3f74ccd8) and we are [using it from Run](https://github.com/ninia/jep/blob/b5ec15ba347b13a58c11ef95fb746635791a52c9/src/main/java/jep/Run.java#L71).  I don't want to suppress that warning because Run is currently our sample code and test case for interactive Jep so I think it should only use non-deprecated API that we plan to support.

Run was recently switched to using SharedInterpreters(To fix #198) but there is currently no non-deprecated way to make a SharedInterpreter interactive. This might be resolved with #168 which wants to specify a JepConfig for SharedInterpreter but the current proposal is to have a single static JepConfig and there is no reason that all SharedInterpreters should be interactive or not, it should be possible to mix. Other options on JepConfig don't make sense for SharedInterpreter so it might need to be split apart anyway which is where we end up writing some new API.

Personally, I would like to drop support for interactive mode entirely in 4.0. The only thing using interactive mode is Jep.eval() and that method has always seemed awkward to me for several reason
1. It is named the same as the Python builtin [eval()](https://docs.python.org/3/library/functions.html#eval) function but it behaves differently, for example it returns a boolean instead of an object.
2. It causes unexpected output, see #151
3. The return value is almost always ignored and completely unnecessary in non-interactive mode.
4. The extra compile step is often unnecessary, for example when we call it internally it is always a complete statement(We should probably switch these to exec anyway for this very reason).
5. Interactive eval has subtle differences from a real Python interpreter, for example typing `bad'` in a python prompt results in an immediate error but the jep interpreter waits for another line of input before pointing out the Syntax Error.

It's entirely possible to write an interactive interpreter, like the Run class without supporting an interactive mode in Jep. Everything that `Jep.eval()` does using cpython is possible using python builtin functions like [compile()](https://docs.python.org/3/library/functions.html#compile) and [exec()](https://docs.python.org/3/library/functions.html#exec). To demonstrate I created a variant of console.py that works with a non-interactive Jep(https://github.com/bsteffensmeier/jep/commit/e4ff78b0a72df3ecfb8490b4afdd9e43180d8ce4), so if we decide to drop interactive mode we could just merge this change to fix the deprecation in Run.java. For reference, in case that link rots, here is the code I used to replace `Jep.eval()`
```
evalLines = []

def jepeval(line):
    global evalLines
    if not line:
        if evalLines:
            code = "\n".join(evalLines)
            evalLines = None
            exec(compile(code, '<stdin>', 'single'), globals(), globals())
        return True
    elif not evalLines:
        try:
            exec(compile(line, '<stdin>', 'single'), globals(), globals())
            return True
        except SyntaxError as err:
            evalLines = [line]
            return False
    else:
        evalLines = evalLines.append(line)
        return False
```

**Comment by ndjensen** [comment]
I suspect but cannot prove that interactive mode was introduced by @mrj0 to support running the jep command line console, similar to running python without args on the command line.  You can see it used in the jep bash script.

I agree that interactive should be removed eventually, based on your bulleted list above.  There's really not a reason for most applications to support interactive mode.  It's really only for an application expecting user input, like the command line console which you have provided an alternative for, or an IDE that wants an interactive Jep interpreter.  By removing the boolean interactive I don't think we are preventing either of those use cases.

**TL;DR** I agree.

**Comment by ndjensen** [comment]
Added classlist_11.txt and reversed the behavior so if it doesn't find a matching version, it uses 11 as the fallback (so it should be future compatible).

**Comment by bsteffensmeier** [comment]
We are tentatively done with code changes, would like to do a release next week. Here is the final checks that need to happen:

- [ ] [Review javadoc](http://ninia.github.io/jep/javadoc/3.9/) specifically I think we need to make sure everything makes sense with the deprecation of Jep, and make sure to switch things referencing eval() to exec() if it makes sense
- [ ] Run through the tests with -Xcheck:jni
- [ ] Update my python refcounting tests and run them: https://github.com/bsteffensmeier/jep/tree/super_leak_test

And then just so we don't miss any steps, here is what we need to do to actually release:
- [ ] Finalize version number everywhere, like this: 3bcf611
- [ ] Run the maven build and make sure it looks right
- [ ] Make a tag, update master
- [ ] Upload to maven central
- [ ] Upload to pypi
- [ ] Send an email to mailing list.
